modules/graceful/releasereopen/releasereopen.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package releasereopen
+
+import (
+	"errors"
+	"sync"
+)
+
+type ReleaseReopener interface {
+	ReleaseReopen() error
+}
+
+type Manager struct {
+	mu      sync.Mutex
+	counter int64
+
+	releaseReopeners map[int64]ReleaseReopener
+}
+
+func (r *Manager) Register(rr ReleaseReopener) (cancel func()) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.counter++
+	r.releaseReopeners[r.counter] = rr
+
+	return func() {
+		r.mu.Lock()
+		defer r.mu.Unlock()
+
+		delete(r.releaseReopeners, r.counter)
+	}
```

Comment:
This does not look alright. If you register two instances and cancel the first one, this will delete the second instance because `r.counter` was incremented by the second call.

Reply:
Oops, good catch. forgot to write a test for it ....

Now it is: f985c0f277bd2733a59d5e4bc2a5efc45413fe95